sympy/polys/matrices/domainscalar.py

```diff
@@ -0,0 +1,86 @@
+"""
+
+Module for the DomainScalar class.
+
+A DomainScalar represents an element which is in a particular
+Domain. The idea is that the DomainScalar class provides the
+convenience routines for unifying elements with different domains.
+
+It assists in Scalar Multiplication and getitem for DomainMatrix.
+
+"""
+from ..constructor import construct_domain
+
+
+class DomainScalar:
+    r"""
+    docstring
+    """
+
+    def __init__(self, element, domain):
+        self.element = element
+        self.domain = domain
+
+    @classmethod
+    def new(cls, element, domain):
+        return cls(element, domain)
+
+    def __repr__(self):
+        return repr(self.element)
+
+    @classmethod
+    def from_sympy(cls, expr):
+        [domain, [element]] = construct_domain([expr])
+        return cls.new(element, domain)
+
+    def to_domain(self, domain):
+        element = domain.convert_from(self.element, self.domain)
+        return self.new(element, domain)
+
+    def unify(self, other):
+        domain = self.domain.unify(other.domain)
+        return self.to_domain(domain), other.to_domain(domain)
+
+    def __add__(self, other):
+        if not isinstance(other, DomainScalar):
+            return NotImplemented
+        self, other = self.unify(other)
+        return self.new(self.element + other.element, self.domain)
+
+    def __sub__(self, other):
+        if not isinstance(other, DomainScalar):
+            return NotImplemented
+        self, other = self.unify(other)
+        return self.new(self.element - other.element, self.domain)
+
+    def __mul__(self, other):
+        if not isinstance(other, DomainScalar):
+            return NotImplemented
+        self, other = self.unify(other)
+        return self.new(self.element * other.element, self.domain)
+
+    def __floordiv__(self, other):
+        if not isinstance(other, DomainScalar):
+            return NotImplemented
+        self, other = self.unify(other)
+        K = self.domain
+        if not K.is_Field:
+            K = K.get_field()
+        return self.new(K.quo(self.element, other.element), self.domain)
+
+    def __pow__(self, n):
+        if not isinstance(n, int):
+            return NotImplemented
+        if self.element == 0 and n < 0:
+            raise ZeroDivisionError
+        return self.new(self.element**n, self.domain)
+
+    def __pos__(self):
+        if self.element < 0:
+            self.element = - self.element
+        return self.new(self.element, self.domain)
```

Comment:
The `__pos__` method is for unary `+` and should not negate the element. Also this should just defer to the operation on element i.e.:
```
return self.new(+self.element, self.domain)
```

Reply:
This also modifies `self.element` which should not happen. `DomainScalar` should be immutable.

Reply:
In order to make it immutable, either we need to override `__setattr__` or use `__slots__`.
IMO introducing `__setattr__` in the constructor will be preferable.
Kindly share your opinion.
Thanks.

Reply:
We don't need to enforce immutability but we should not mutate the value as this code does.

Reply:
Sure, thanks.